### DISCUSSION TITLE
How to optimize speed for after_build_file？

### DISCUSSION THREAD

**Discussion by Zmmfly** [xmake-io/xmake#4605] [General]
目前用 xmake 改造一个用 armcc 作为编译器的sdk，需要用-D -E生成预处理文件然后用自带的工具处理。
用 `after_build_file` 做到了生成预处理这个功能，但看输出，after_build_file好像是在编译完成后才执行，而且还是顺序执行
翻遍了文档也没看到介绍给 `after_build_file` 加速的

下面是写的after_build_file，可能有点烂哈哈
```lua
after_build_file(function(target, src, opt) 
    import("core.tool.compiler")

    local c_dir        = vformat(path.translate("$(scriptdir)/tools"))
    local c_ppc        = vformat("%s", path.translate(c_dir.."/prepass/Hermon/Release/PPC.exe"))
    local f_obj        = target:objectfile(src)
    local p_target     = target:scriptdir()
    local p_target_obj = p_target..path.translate("/build/obj")
    local f_ppp        = path.translate(p_target_obj.."/"..path.basename(path.basename(f_obj))..".ppp")
    local f_pp         = path.translate(p_target_obj.."/"..path.basename(path.basename(f_obj))..".pp")

    -- mk target/build
    if not os.exists(p_target_obj) then 
        os.mkdir(p_target_obj)
    end

    print("mk pp for"..src)
    compiler.compile(src, f_ppp, {target=target, configs={cflags = "-E -C ", cxxflags="-E -C "}})

    -- convert ppp to pp
    os.run(format("%s %s %s", c_ppc, f_ppp, f_pp))
end)
```

**Comment by waruqi** [comment]
你想怎么优化？它编译就这个速度，不同 obj 之间调度已经是并行编译了，同一 obj ，肯定得先编译完，才能执行 after_build_file 。。顾名思义，after build...

**Reply by Zmmfly** [reply]
我描述得不够清楚，after_build确实是我的理解问题；

这个case其实是每一个文件，除了编译外，还得另外用-E -C再生成一次预处理后的文本文件，需要加速的是这个生成的过程；
这就相当于再走一次编译流程，可以用多线程加速；

原先设想的是能在编译的同时，用xmake内置的多线程同时进行预处理，后来也用on_build_file进行了尝试，效果不是太理想；

索性写了个外部的多线程运行命令，在after_build_file里把预处理的命令写入文本，再在after_build用外部命令进行多线程预处理；

最终，需要的可能是一个内置的多线程模块，这样就无需多一个调用外部命令的过程

**Reply by waruqi** [reply]
还是没懂你想干啥，lua 是没有多线程的，只有协程来处理异步。。目前 xmake 的协程封装，仅自动处理进程，pipe，socket 之间的切换调度，其他的 你只能自己调用 yield 去切换。。
